### ISSUE TITLE
Some header files in the latest master are buggy.

### ISSUE THREAD

**Issue by beef9999** [axboe/liburing#575]
After I upgraded liburing from `469b5e80ce42be53ec748ddea8523a3c16ea5465` (2021-11-17) to the latest master, my old program got a Segmentation fault when trying to use the data pointer from `io_uring_cqe_get_data(cqe)`.

Not the libs, but the changed header files caused such core dumps.

**Comment by ammarfaizi2** [comment]
> After I upgraded liburing from `469b5e80ce42be53ec748ddea8523a3c16ea5465` (2021-11-17) to the latest master, my old program got a Segmentation fault when trying to use the data pointer from `io_uring_cqe_get_data(cqe)`.
> 
> Not the libs, but the changed header files caused such core dumps.

Do you have a simplified reproducer we can try?

**Comment by beef9999** [comment]
@ammarfaizi2 At first I was trying to linking to an old and new liburing.so seperately to demonstrate the bug, but then I found that it was the header files causing the problems. It happened in compiling time.

Because my project is an internal project and compiling is very sophisticated, so I'm not enable to prove it, at least from your perspective.

**Comment by ammarfaizi2** [comment]
I am unable to reproduce this. Can you create a simple program that provokes the bug and post it here?

**Comment by beef9999** [comment]
@ammarfaizi2 I'm afraid our code is not open sourced, even running a simple program would require to refer to our base framework. Maybe you can send me a email (beef9999@qq.com), so I can reply to you with the binary (built in DEBUG mode).

**Comment by ammarfaizi2** [comment]
> @ammarfaizi2 I'm afraid our code is not open sourced, even running a simple program would require to refer to our base framework. Maybe you can send me a email ([beef9999@qq.com](mailto:beef9999@qq.com)), so I can reply to you with the binary (built in DEBUG mode).

Sent.

**Comment by CarterLi** [comment]
You'd better upgrade the libs accordingly, as they may not be ABI compatible.

**Comment by axboe** [comment]
Yes, re-compiling against a new header but using the old lib is a recipe for disaster. The new library should be ABI compatible, hence upgrades should be fine, but do not compile with the 2.2-git header and the 2.1 lib. That's not a supported setup.

**Comment by beef9999** [comment]
No I didn't mean it. I was just trying to link the program to .so to demonstrate the bug.

In fact we were linking the .a lib, which means both header files and libs have been updated to the master version.

The network test case that used to be passed would core dump for now. The io test case will sometimes return 0 when doing io_uring read/write.

**Comment by ammarfaizi2** [comment]
@beef9999 I got your email and tried to run both binary files, but I didn't get any segfault. If you can easily reproduce it from your end, can you do `git bisect` to find the first bad commit?

Just looking at the binary files doesn't help. I can't debug it.

**Comment by beef9999** [comment]
```
# git bisect bad
8ecd3fd959634df81d66af8b3a69c16202a014e8 is the first bad commit
commit 8ecd3fd959634df81d66af8b3a69c16202a014e8
Author: Jens Axboe <axboe@kernel.dk>
Date:   Wed Dec 22 19:55:20 2021 -0700

    Don't clear sqe->user_data as part of command prep

    Reasoning:

    1) If the app doesn't set it, it should not be reading it.
    2) I've gotten more than one report on someone setting user_data
       before calling io_uring_prep_cmdtype() and it's a bit confusing.

    I think we'll be fine not clearing this, and it may actually help
    some cases that end up being a bit confusing, particular for people
    new to the interface.

    Signed-off-by: Jens Axboe <axboe@kernel.dk>
```

This change might be friendly to new users, but caused troubles to the old...

My old program regards user_data as zero by default. It's vulnerable to random values.

**Comment by YoSTEALTH** [comment]
> It's vulnerable to random values.

Yes, since new `sqe` space is reused from the old submitted `sqe`, there is a chance if you don't set the `user_data=0` it can use the older `user_data` value.

Maybe it should be set here:
https://github.com/axboe/liburing/blob/master/src/include/liburing.h#L918

```c
sqe = &sq->sqes[sq->sqe_tail & *sq->kring_mask];
sqe->user_data = 0;
```

or `memset` it?

**Comment by axboe** [comment]
OK, that does make sense. You can obviously argue that this is a regression or change of behavior, which it is, but it's at least trivial to resolve (always set ->user_data if you always use it at completion time). I'm reluctant to revert this change.
